cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -41,291 +41,719 @@
 
-  //! @brief Forwards the stateful properties
-  _CCCL_TEMPLATE(class _Property)
-  _CCCL_REQUIRES(property_with_value<_Property> _CCCL_AND(_CUDA_VSTD::__is_included_in_v<_Property, _Properties...>))
-  _CCCL_NODISCARD_FRIEND __property_value_t<_Property> get_property(const basic_any_resource& __res, _Property) noexcept
+template <class _Derived>
+struct __with_try_get_property
+{
+  template <class _Property>
+  _CUDAX_HOST_API _CCCL_NODISCARD_FRIEND auto
+  try_get_property(const _Derived& __self, _Property) noexcept -> __try_property_result_t<_Property>
   {
-    _CUDA_VMR::_Property_vtable<_Property> const& __prop = __res;
-    return __prop.__property_fn(__res._Get_object());
+    auto __prop = __cudax::dynamic_any_cast<const __iproperty<_Property>*>(&__self);
+    if constexpr (_CUDA_VSTD::is_same_v<__property_result_t<_Property>, void>)
+    {
+      return __prop != nullptr;
+    }
+    else if (__prop)
+    {
+      return get_property(*__prop, _Property{});
+    }
+    else
+    {
+      return _CUDA_VSTD::nullopt;
+    }
   }
 };
 
-//! @brief Checks whether a passed in type is a specialization of basic_any_resource
-template <class _Ty, _CUDA_VMR::_AllocType _Alloc_type, class... _Properties>
-_CCCL_INLINE_VAR constexpr bool __is_basic_any_resource<_Ty, basic_any_resource<_Alloc_type, _Properties...>> = true;
+// `any_resource` wraps any given resource that satisfies the required
+// properties. It owns the contained resource, taking care of construction /
+// destruction. This makes it especially suited for use in e.g. container types
+// that need to ensure that the lifetime of the container exceeds the lifetime
+// of the memory resource used to allocate the storage
+template <class... _Properties>
+struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES any_resource
+    : basic_any<__iresource<_Properties...>>
+    , __with_try_get_property<any_resource<_Properties...>>
+{
+  using any_resource::basic_any::basic_any;
+
+private:
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
+                "The properties of cuda::experimental::any_resource must contain at least one execution space "
+                "property!");
+  using any_resource::basic_any::interface;
+};
+
+// ``any_async_resource`` wraps any given async_resource that satisfies the
+// required properties. It owns the contained resource, taking care of
+// construction / destruction. This makes it especially suited for use in e.g.
+// container types that need to ensure that the lifetime of the container
+// exceeds the lifetime of the memory resource used to allocate the storage
+template <class... _Properties>
+struct _LIBCUDACXX_DECLSPEC_EMPTY_BASES any_async_resource
+    : basic_any<__iasync_resource<_Properties...>>
+    , __with_try_get_property<any_async_resource<_Properties...>>
+{
+  using any_async_resource::basic_any::basic_any;
+
+private:
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
```

Comment:
I was wondering since the sliced off properties are now accessible with the `try_get_property` if we could drop this requirement?

Reply:
i lean towards "no". @miscco?